### PR TITLE
8299570: [JVMCI] Insufficient error handling when CodeBuffer is exhausted

### PR DESCRIPTION
**Pull request by dougxc**
This PR fixes the handling of a full CodeBuffer when emitting stubs as part of JVMCI code installation.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299570](https://bugs.openjdk.org/browse/JDK-8299570): [JVMCI] Insufficient error handling when CodeBuffer is exhausted

### Reviewers
 * [Tom Rodriguez](https://openjdk.org/census#never) (@tkrodriguez - **Reviewer**) ⚠️ Review applies to [7f1d45aa](https://git.openjdk.org/jdk/pull/11945/files/7f1d45aa0b20e89faf386430de472d3dd2369d3e)
 * [Andrew Dinn](https://openjdk.org/census#adinn) (@adinn - **Reviewer**) ⚠️ Review applies to [9b1d1fbe](https://git.openjdk.org/jdk/pull/11945/files/9b1d1fbe1aa396b1dba811c995a0b473495e773b)
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11945/head:pull/11945` \
`$ git checkout pull/11945`

Update a local copy of the PR: \
`$ git checkout pull/11945` \
`$ git pull https://git.openjdk.org/jdk pull/11945/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11945`

View PR using the GUI difftool: \
`$ git pr show -t 11945`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11945.diff">https://git.openjdk.org/jdk/pull/11945.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp
Comment (tkrodriguez) [inline src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp:539]: The if (stub == null) test below should be the else branch of this if which I think makes it clearer.  Why do we even bother returning the stub?

File: src/hotspot/share/jvmci/jvmciCodeInstaller.cpp
Comment (tkrodriguez) [inline src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:1181]: We use estimate_stubs_size to presize the buffer to include space for these so it should never fail I think, but we should be checking anyway.  Maybe estimate_stubs_size should be computing space for the aarch64 trampolines?

**Review by tkrodriguez — APPROVED** [review]

File: src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp
Comment (tkrodriguez) [inline src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp:542]: The if (stub == null) test below should be the else branch of this if which I think makes it clearer. Why do we even bother returning the stub?

Comment (dougxc) [reply src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp:542]: >The if (stub == null) test below should be the else branch of this if which I think makes it clearer

I'm suspecting this is indeed a dormant bug. I will investigate further.

@theRealAph any chance you recall the intended logic here?

File: src/hotspot/share/jvmci/jvmciCodeInstaller.cpp
Comment (teshull) [reply src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:1181]: it looks like estimate_stubs_size is computing space for the trampolines via 

`size += trampoline_stubs * CompiledStaticCall::to_trampoline_stub_size();`?

Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:1181]: It seems like the estimate agrees with the code emitted for the trampoline, so we either have a general problem in this computation or we have some other unexpected use of code cache space.  It isn't critical that we get this estimate right but we do try to avoid resizing during code installation as it can be expensive.

Comment (dougxc) [reply src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:1181]: Yes, this is a very defensive (and cheap) check that should never fail. Still, better to have it.

File: src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp
Comment (dougxc) [reply src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp]: I've pushed https://github.com/openjdk/jdk/pull/11945/commits/0e4fb6501dbff8259c7ae131bf81e03e7e362b94 which attempts to clear up the logic in this method. It would be great if some aarch64 experts could help review it.

Comment (adinn) [reply src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp]: @dougxc I'm not sure why the above code is written the way it is rather than the way you rewrote it. I cannot see any reason why there should already be a trampoline stub in place when trampoline_jump is called given how it is being called at present. I thought perhaps it might be something to do with the (newly introduced) shared trampoline code but that is not relevant here and, besides, this routine has been thew way it is since it was first introduced. 

I know the trampoline (and related far jump) code has been subject to change over the years so it may be something to do with how this routine was called in an earlier incarnation of the code. Andrew Haley (@aph) will have a better idea than me as he was the original author.

Anyway, if we may need far branches and the call to is_NativeCallTrampolineStub_at fails then it does not seem tome to make any sense to call set_destination (at least null is returned which is correct). So, I think your rewrite looks like it is doing the right thing.

I think you probably need an ok from Andrew Haley here though.

Comment (dougxc) [reply src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp]: Ok, thanks for your input. I'll wait for @theRealAph to review it as well.

Comment (dougxc) [reply src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp]: @theRealAph @adinn can I now merge this PR?

Comment (adinn) [reply src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp]: @dougxc Still ok with me. I just pinged Andrew Haley to see if he is ok with it.

**Review by adinn — APPROVED** [review]
I'm happy to approve this.

**Comment by adinn** [comment]
@dougxc I'm assuming the test failures are unrelated? If so then ok to push.

File: src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp:548]: This is all very complicated.  Can't we just add `JVMCI_ERROR`s where we need them?

Comment (dougxc) [reply src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp:548]: I'm not following - isn't this exactly what the code is doing? Maybe you could demonstrate how you think it should look.

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp:548]: Maybe it would be nicer to get the `! far_branches` code path out of the way first, and return immediately.

Comment (dougxc) [reply src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp:548]: Something like this?
```
diff --git a/src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp b/src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp
index 88dc59f80d0..83ec182d2c7 100644
--- a/src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp
+++ b/src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp
@@ -532,21 +532,22 @@ void NativeCallTrampolineStub::set_destination(address new_destination) {
 void NativeCall::trampoline_jump(CodeBuffer &cbuf, address dest, JVMCI_TRAPS) {
   MacroAssembler a(&cbuf);
 
-  if (a.far_branches()) {
-    if (!is_NativeCallTrampolineStub_at(instruction_address() + displacement())) {
-      address stub = a.emit_trampoline_stub(instruction_address() - cbuf.insts()->start(), dest);
-      if (stub == nullptr) {
-        JVMCI_ERROR("could not emit trampoline stub - code cache is full");
-      }
-      // The relocation is created while emitting the stub will ensure this
-      // call instruction is subsequently patched to call the stub.
-    } else {
-      // Not sure how this can be happen but be defensive
-      JVMCI_ERROR("single-use stub should not exist");
-    }
-  } else {
+  if (!a.far_branches()) {
     // If not using far branches, patch this call directly to dest.
     set_destination(dest);
+    return;
+  }
+
+  if (!is_NativeCallTrampolineStub_at(instruction_address() + displacement())) {
+    address stub = a.emit_trampoline_stub(instruction_address() - cbuf.insts()->start(), dest);
+    if (stub == nullptr) {
+      JVMCI_ERROR("could not emit trampoline stub - code cache is full");
+    }
+    // The relocation is created while emitting the stub will ensure this
+    // call instruction is subsequently patched to call the stub.
+  } else {
+    // Not sure how this can be happen but be defensive
+    JVMCI_ERROR("single-use stub should not exist");
   }
 }
 #endif
```

Comment (dougxc) [reply src/hotspot/cpu/aarch64/nativeInst_aarch64.cpp:536]: Done.

**Comment by dougxc** [comment]
Thanks for the reviews @adinn @theRealAph and @tkrodriguez .

/integrate

**Pull request closed without merging**
